src/Peachpie.Library/Ftp.cs

```diff
@@ -875,5 +881,421 @@ public static bool ftp_site(PhpResource ftp_stream, string command)
+        /// <param name="mode">The transfer mode. Must be either FTP_ASCII or FTP_BINARY.</param>
+        /// <param name="resumepos">The position in the remote file to start downloading from.</param>
+        /// <returns>Returns FTP_FAILED or FTP_FINISHED or FTP_MOREDATA.</returns>
+        public static int ftp_nb_get(Context ctx, PhpResource ftp_stream, string local_file, string remote_file, int mode = FTP_BINARY, int resumepos = 0)
+        {
+            // Check ftp_stream resource
+            var resource = ValidateFtpResource(ftp_stream);
+            if (resource == null)
+            {
+                return FTP_FAILED;
+            }
+
+            if (resumepos != 0) // There is no API for this parameter in FluentFTP Library.
+                PhpException.ArgumentValueNotSupported(nameof(resumepos), resumepos);
+
+            resource.PrepareTaskFunction(resource, mode);
+
+            resource.TaskFunction = resource.Client.DownloadFileAsync(local_file, remote_file, FtpLocalExists.Overwrite, FtpVerify.None, null, resource.TokenSource.Token);
+
+            return TasksGetInfo(resource);
+        }
+
+        /// <summary>
+        /// Retrieves a file from the FTP server and writes it to a local file (non-blocking).
+        /// </summary>
+        /// <param name="ftp_stream">The link identifier of the FTP connection.</param>
+        /// <param name="handle">An open file pointer in which we store the data.</param>
+        /// <param name="remote_file">The remote file path.</param>
+        /// <param name="mode">The transfer mode. Must be either FTP_ASCII or FTP_BINARY.</param>
+        /// <param name="resumepos">The position in the remote file to start downloading from.</param>
+        /// <returns>Returns FTP_FAILED or FTP_FINISHED or FTP_MOREDATA.</returns>
+        public static int ftp_nb_fget(PhpResource ftp_stream, PhpResource handle, string remote_file, int mode = FTP_BINARY, int resumepos = 0)
+        {
+            // Check file resource
+            var stream = PhpStream.GetValid(handle);
+            if (stream == null)
+            {
+                return FTP_FAILED;
+            }
+
+            // Check ftp_stream resource
+            var resource = ValidateFtpResource(ftp_stream);
+            if (resource == null)
+            {
+                return FTP_FAILED;
+            }
+
+            // Ignore autoresume if autoseek is switched off 
+            if (resource.Autoseek && resumepos == FTP_AUTORESUME)
+            {
+                resumepos = 0;
+            }
+
+            if (resource.Autoseek && resumepos != 0)
+            {
+                if (resumepos == FTP_AUTORESUME)
+                {
+                    stream.Seek(0, SeekOrigin.End);
+                    resumepos = stream.Tell();
+                }
+                else
+                {
+                    stream.Seek(resumepos, SeekOrigin.Begin);
+                }
+            }
```

Comment:
It's the same logic as in `ftp_fget`, wouldn't it be better to move it to a common helper method? Not only for code size, but also for maintainability (e.g. if we need to change the logic, we can do it in one place).